File: src/cpp/multithreading/03_multiple_worker_threads.cpp
#include <iostream>
#include <thread>
#include <vector>

void foo(int id) {
    std::cout << "Function foo() starts, Thread id: " << id << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(1));
    std::cout << "Function foo() finishes, Thread id: " << id << std::endl;
}

void bar(int id) {
    std::cout << "Function bar() starts, Thread id: " << id << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    std::cout << "Function bar() finishes, Thread id: " << id << std::endl;
}

int main() {
    std::cout << "Main thread starts" << std::endl;

    std::vector<std::thread> threads;

    for (int i = 0; i < 5; ++i) {
        threads.emplace_back(foo, i);
    }

    for (int i = 0; i < 5; ++i) {
        threads.emplace_back(bar, i);
    }

    for (auto &thread : threads) {
        thread.join();
    }

    std::cout << "Main thread finishes" << std::endl;

    return 0;
}

File: src/python/multithreading/03_multiple_worker_threads.py
"""
This script demonstrates the creation of multiple threads that execute different functions concurrently.
It creates two sets of threads, one set running the 'foo' function and the other set running the 'bar' function.
Each function has a different execution time, and each thread is given an ID to track its progress.
"""

import time
from threading import Thread


def foo(id):
    print(f"Function foo() starts, Thread id: {id}")
    time.sleep(1)
    print(f"Function foo() finishes, Thread id: {id}")


def bar(id):
    print(f"Function bar() starts, Thread id: {id}")
    time.sleep(0.1)
    print(f"Function bar() finishes, Thread id: {id}")


if __name__ == "__main__":
    print("Main thread starts")

    threads = []
    for i in range(5):
        threads.append(Thread(target=foo, args=(i,)))

    for i in range(5):
        threads.append(Thread(target=bar, args=(i,)))

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    print("Main thread finishes")
